arrow-core-data/src/main/kotlin/arrow/core/SuspendingMonadContinuation.kt

```diff
@@ -0,0 +1,81 @@
+package arrow.core
+
+import arrow.Kind
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.loop
+import java.lang.RuntimeException
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.resumeWithException
+
+internal const val UNDECIDED = 0
+internal const val SUSPENDED = 1
+
+@Suppress("UNCHECKED_CAST")
+abstract class SuspendMonadContinuation<F, A>(private val parent: Continuation<Kind<F, A>>) : Continuation<Kind<F, A>> {
+
+  class ShortCircuit(val e: Any?) : Throwable(message = null, cause = null) {
+    override fun fillInStackTrace(): Throwable = this
+  }
+
+  abstract operator fun <A> Kind<F, A>.not(): A
+
+  abstract fun ShortCircuit.recover(): Kind<F, A>
+
+  /**
+   * State is either
+   *  0 - UNDECIDED
+   *  1 - SUSPENDED
+   *  Any? (3) `resumeWith` always stores it upon UNDECIDED, and `getResult` can atomically get it.
+   */
+  private val _decision = atomic<Any>(UNDECIDED)
+
+  override val context: CoroutineContext = EmptyCoroutineContext
+
+  override fun resumeWith(result: Result<Kind<F, A>>) {
+    _decision.loop { decision ->
+      when (decision) {
+        UNDECIDED -> {
+          val r: Kind<F, A>? = when {
+            result.isFailure -> {
+              val e = result.exceptionOrNull()
+              if (e is ShortCircuit) e.recover() else null
```

Comment:
If you want to do a one-liner

```suggestion
              (result.exceptionOrNull() as? ShortCircuit)?.recover()
```